staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -811,66 +721,51 @@ func (c *Cacher) dispatchEvents() {
 func (c *Cacher) dispatchEvent(event *watchCacheEvent) {
 	c.startDispatching(event)
 	defer c.finishDispatching()
-	// Watchers stopped after startDispatching will be delayed to finishDispatching,
 
-	// Since add() can block, we explicitly add when cacher is unlocked.
-	// Dispatching event in nonblocking way first, which make faster watchers
-	// not be blocked by slower ones.
-	if event.Type == watch.Bookmark {
-		for _, watcher := range c.watchersBuffer {
-			watcher.nonblockingAdd(event)
-		}
-	} else {
-		c.blockedWatchers = c.blockedWatchers[:0]
-		for _, watcher := range c.watchersBuffer {
-			if !watcher.nonblockingAdd(event) {
-				c.blockedWatchers = append(c.blockedWatchers, watcher)
-			}
+	// dispatch BOOKMARK event firstly
+	bookmarkEvent := &watchCacheEvent{
+		Type:            watch.Bookmark,
+		Object:          c.newFunc(),
+		ResourceVersion: event.ResourceVersion,
+	}
+	for _, watcher := range c.watchersBookmarkBuffer {
```

Comment:
It means that for every event, you're also trying to dispatch bookmark for every watcher.
I know that it doesn't mean that the event is sent to the client (just cached in the handler), but that's hell a lot of work - did you run any large-scale tests of it? I'm fairly sure this will cause issues for high-churn resources...

Reply:
I do not think it's a problem. It just put an object to a channel and it's non-blocking. The only problem I think is the receiver(watch rest handler) need to decode the event, but I think I can fix it.

Reply:
Think about the older way, we need to iterate a map which the key is a UNIX timestamp!

Reply:
The older codes are really tricky and make me confused.

Reply:
> It means that for every event, you're also trying to dispatch bookmark for every watcher.

No, I just try to dispatch bookmark for the watchers which does not trigger the event. Please see the following codes:

```go
		for _, triggerValue := range triggerValues {
			for _, watcher := range c.watchers.valueWatchers[triggerValue] {				for v, watchers := range c.watchers.valueWatchers {
				c.watchersBuffer = append(c.watchersBuffer, watcher)					if triggerValue == v {
					for _, watcher := range watchers {
						c.watchersBuffer = append(c.watchersBuffer, watcher)
					}
					continue
				}

 				// here we need to send BOOKMARK event
				for _, watcher := range watchers {
					c.watchersBookmarkBuffer = append(c.watchersBookmarkBuffer, watcher)
				}
```

Reply:
The `c.watchersBookmarkBuffer` only adds the watchers which does not trigger the event

Reply:
> I do not think it's a problem. It just put an object to a channel and it's non-blocking. 

Putting object to a channel is NOT free. It's actually quite expensive, requires locking etc.

There is significant difference with this approach.
Let's say you have 5k-node cluster.
So there are basically O(5k) watchers of pods, but for any given pod, there are basically only O(1) watchers interested in that change.

With current code, we still put O(1) events into channels, with your code you put O(5k).
That's different order of magnitude of operations to do.

Before I do a deep review and let this in, I need to see a convincing performance data that this approach is indeed better.

Reply:
If dispatch every bookmark event costs too much，there is another way：
we can record the latest RV and then use a ticker to dispatch bookmark event in every 2s just like before.

Reply:
> With current code, we still put O(1) events into channels, with your code you put O(5k).
That's different order of magnitude of operations to do.

With current code, we need to iterate an UNIX timestamp key map.

```go
func (t *watcherBookmarkTimeBuckets) popExpiredWatchers() [][]*cacheWatcher {	
	currentBucketID := t.clock.Now().Unix()	
	// There should be one or two elements in almost all cases	
	expiredWatchers := make([][]*cacheWatcher, 0, 2)	
	t.lock.Lock()	
	defer t.lock.Unlock()	
	for ; t.startBucketID <= currentBucketID; t.startBucketID++ {	
		if watchers, ok := t.watchersBuckets[t.startBucketID]; ok {	
			delete(t.watchersBuckets, t.startBucketID)	
			expiredWatchers = append(expiredWatchers, watchers)	
		}	
	}	
	return expiredWatchers	
}
```

Reply:
> If dispatch every bookmark event costs too much，there is another way：
we can record the latest RV and then use a ticker to dispatch bookmark event in every 2s just like before.

How it would be different from what we're currently doing.
I don't see the gain in refactoring.